Fuzzy watchface

dyamon opened the pull request:
This is a first attempt at creating a fuzzy watchface similar to the one present on the Pebble and requested in #653 .

Feedback on the code is highly appreciated since I'm more of a C person and I'm not familiar with the C++ coding style. The code needs at least to be cleaned from some personal notes.

Also feedback on the design itself is appreciated since I don't own a Pebble and I partially based the design on some videos online.

At the moment this is the result:

<p align="center">
    <img src="https://user-images.githubusercontent.com/6771224/132925529-a86c38aa-2dd9-47b3-b914-0df9677259dc.png" width="500" >
</p>

I kinda like the bottom left alignment but we can change this quite easily. Also the font has been resized after the shots to better accommodate longer text.

Finally I have a slightly different design in mind, but I want to upload a mockup before going back to coding.

Closes #653

schneidr commented:
I love it. I just started to work on exactly this watchface, but I didn't get very far yet.
One thing I have to criticize though, it seems your watchface shows every exact minute, which is not exactly "fuzzy". The original Pebble watchface (I've been using it for years) only showed 5 minute intervals.

dyamon commented:
Oooh ok, didn't know that! This makes sense, and might make things easier for internationalisaion later on.

I guess it doesn't really matter whether you round up or down but, is this what's happening?
```
8:00-8:04 -> "eight o'clock"
8:05-8:09 -> "five past eight"
8:10-8:14 -> "ten past eight"
8:15-8:19 -> "quarter past eight"
8:20-8:24 -> "twenty past eight"
8:25-8:29 -> "twenty five past eight"
8:30-8:34 -> "half past eight"
8:35-8:39 -> "twenty five to nine"
8:40-8:44 -> "twenty to nine"
8:45-8:49 -> "quarter to nine
8:50-8:54 -> "ten to nine"
8:55-8:59 -> "five to nine"
```

Also what about hours? It sticks to 12H format right? Do 00:00 and 12:00 turn into "midnight" and "midday" respectively?

adocampo commented:
I love it! I would rather prefer a centered position, but I can ge used to bottom left as well.
What I really would like to see is some kind of imprecision, so when the minutes are less than 3 it rounds to 0 and when it's between minute 4 and 7 rounds to 5.

> Oooh ok, didn't know that! This makes sense, and might make things easier for internationalisaion later on.
> 
> I guess it doesn't really matter whether you round up or down but, is this what's happening?
> 
> ```
> 8:00-8:04 -> "eight o'clock"
> 8:05-8:09 -> "five past eight"
> 8:10-8:14 -> "ten past eight"
> 8:15-8:19 -> "quarter past eight"
> 8:20-8:24 -> "twenty past eight"
> 8:25-8:29 -> "twenty five past eight"
> 8:30-8:34 -> "half past eight"
> 8:35-8:39 -> "twenty five to nine"
> 8:40-8:44 -> "twenty to nine"
> 8:45-8:49 -> "quarter to nine
> 8:50-8:54 -> "ten to nine"
> 8:55-8:59 -> "five to nine"
> ```
> 
> Also what about hours? It sticks to 12H format right? Do 00:00 and 12:00 turn into "midnight" and "midday" respectively?

I was typing this message when yours appeared on github, so I'm quoting it.

I love it! I would rather prefer a centered position, but I can get used to bottom left as well.
What I really would like to see is some kind of imprecision, so when the minutes are from 0 to 3 it rounds to 0 and when it's between minute 4 and 7 rounds to 5. So, for example:
```
8:02 -> "eight o'clock'
8:04 -> "five past eight"
8:09 -> "ten past eight"
8:14 -> "quarter past eight"
8:17 -> "quarter past eight"
8:18 -> "twenty  past eight"
```

Regarding 00h and 12h, I would see more normal to say "twelve o'clock" rather than "midnight" or "midday", that kind of imprecision would be even more fuzzy (00:01h-11:59 → morning, 12:00h → midday/afternoon, 12:01-17h → afternoon, 17:01-20h → evening, 20:01-23:59 → night, 00h → midnight) and I wouldn't include those terms on a wristwatch. (KDE Plasma's fuzzy clock has those levels - and even more - of fuzziness, but on a wristwatch, I would see more logical to have an imprecise fuzzy watch to say the time on 0 and 5 minutes, like your example but with my suggestion.

As sidenote, pebble fuzzy watchface was able to show the exact digital time during a few seconds by shaking the wrist, so if you need for some reason know the exact time, you were able to do it

schneidr commented:
> Oooh ok, didn't know that! This makes sense, and might make things easier for internationalisaion later on.
> 
> I guess it doesn't really matter whether you round up or down but, is this what's happening?
> 
> ```
> 8:00-8:04 -> "eight o'clock"
> 8:05-8:09 -> "five past eight"
> 8:10-8:14 -> "ten past eight"
> 8:15-8:19 -> "quarter past eight"
> 8:20-8:24 -> "twenty past eight"
> 8:25-8:29 -> "twenty five past eight"
> 8:30-8:34 -> "half past eight"
> 8:35-8:39 -> "twenty five to nine"
> 8:40-8:44 -> "twenty to nine"
> 8:45-8:49 -> "quarter to nine
> 8:50-8:54 -> "ten to nine"
> 8:55-8:59 -> "five to nine"
> ```
> 
> Also what about hours? It sticks to 12H format right? Do 00:00 and 12:00 turn into "midnight" and "midday" respectively?

I always implemented it with +-2 minutes. 
07:58 - 08:02 -> 8 o'clock
08:03 - 08:07 -> five after eight
This fits pretty nicely. And it stays with 12 hour format, yes, just like one would say it.

Riksu9000 reviewed (changes requested):
The 42px font we currently have can fit 9.5 characters on a line. If it were 40px it would fit 10 and 36px can fit 11. Would it make sense to have just a 40px font for everything, at least for now?

>Also, this is what I also had in mind for an alternative design:

I prefer the original design on first glance.

>I'm still not completely happy, as I would like the text to "scroll" when the time changes.

In theory it might look good, but the screen refresh is actually quite slow. Scrolling large text probably ends up looking too choppy. The reason screen scrolling looks fine is because it uses hardware scrolling.

dyamon commented on src/displayapp/screens/WatchFaceFuzzy.cpp:
```diff
@@ -0,0 +1,120 @@
+  /* Why would we need this? */
+  //settingsController.SetClockFace(0);
+
+  /*
+   * `lv_scr_act()` get the active screen on the default display
+   * (note that we only have one display in this setup, but LVGL has
+   * multidisplay support as well).
+   *
+   * `nullptr` is used simply because we don't want to *copy* the label
+   * from any other object.
+   */
```
Yeah, as I said these were personal notes, since I'm not completely familiar with LVGL. They'll be gone before turning this into a proper PR.

dyamon commented on src/displayapp/screens/WatchFaceFuzzy.cpp:
```diff
@@ -0,0 +1,120 @@
+   * object size. In this case we want to break the text to keep the
+   * width and extend the height of the obj.
+   *
+   * Note: this doesn't work properly or it is someone dependent on when
+   * it is called. Avoiding it for now and specifying linebreaks
+   * directly in the string.
+   */
+  //lv_label_set_long_mode(label_time, LV_LABEL_LONG_BREAK);
```
I saw the docs and tried a few combinations a couple of days ago, but no such luck. I still think manual breaks work better.

Riksu9000 replied:
I used long mode in #489, so it should work fine. It probably makes sense to use manual brakes in some places, but I think having long mode working as a backup is important to make it less likely to break, especially if we have more languages.

dyamon replied:
So, I managed to make "long mode" work but I'm seeing a couple of problems:
- It breaks words;
- what we want is to "substitute" a space with a `\n`, but this mode inserts a `\n` between chars. This creates a weird alignment.

Riksu9000 replied:
It shouldn't break words. It doesn't in #489 either. Unless maybe if a word is too long to fit on a single line anyway. Did you set the label width correctly?
https://github.com/JF002/InfiniTime/blob/c9aeef94ea824ad7304ca3f0bd76a476f4744ff4/src/libs/lv_conf.h#L498

dyamon replied:
So, in the end, "long mode" works but it is not completely compatible with the recoloring feature. I'm keeping it on as a safety net, but also include hard breaks in the text for better control over the displayed output.

dyamon commented on src/displayapp/screens/WatchFaceFuzzy.h:
```diff
@@ -0,0 +1,28 @@
+//#include "ScreenList.h"
+#include "components/datetime/DateTimeController.h"
+
+namespace Pinetime {
+    namespace Applications {
+        namespace Screens {
+            class WatchFaceFuzzy : public Screen {
+                public:
+                WatchFaceFuzzy(DisplayApp* app, Controllers::DateTime& dateTimeController);
```
Yep, as for the personal notes, I'll run it in the end.

dyamon commented on src/displayapp/screens/WatchFaceFuzzy.cpp:
```diff
@@ -0,0 +1,120 @@
+  } else {
+    lv_label_set_text_fmt(label_time, "twenty\n%s %s\n#%s %s#", strings[minutesUnits - 1], pastTo, hoursAccent, hoursStr);
+  }
+  /* Align the label w.r.t. another object (active screen in this case)
+   * You can use the last two parameters to move the obj around 
+   * *after the alignment*.
+   * See https://docs.lvgl.io/latest/en/html/_images/align.png for all
+   * possible alignments.
+   *
+   * NOTE: you should set the alignment after determining the size
+   * (and content) of the label to get a reliable result!
+   */
+  lv_obj_align(label_time, lv_scr_act(), LV_ALIGN_IN_BOTTOM_LEFT, 10, -10);
```
I do not call `lv_obj_align()` in the constructor because you need to set the text before doing any alignment. I'll comment below about the off screen issue.

Riksu9000 replied:
Ah I confused label_align with obj_align. My bad.

dyamon commented:
> The 42px font we currently have can fit 9.5 characters on a line. If it were 40px it would fit 10 and 36px can fit 11. Would it make sense to have just a 40px font for everything, at least for now?

So now that the clock will be fuzzy, the longest string we deal with is "five past" which is 9 chars and should perfectly fit with the 42px font and leave some space on the borders as well. Still the 42px font only contains digits atm so we need to add lowercase letters anyway.

Thanks for the info about animations!

adocampo commented:
> ![fuzzy-alt](https://user-images.githubusercontent.com/6771224/132942204-83f59354-4ca5-44af-b18a-030661a4a352.png)
> 
I like more those colors than on the first mockup, and because the text is more centerd, but I'd rather prefer a black background: it is more discrete and perhaps more battery saving (no idea about that).
> I'm still not completely happy, as I would like the text to "scroll" when the time changes. If you guys like the concept I will dive into the topic of animations in LVGL. Also note that the two designs are not mutually exclusive. If you make colors customisable (like for the pinetime style) one could always set the bg text to black and forget it even exists.
> 
I have my PT screen force to sleep after 5s, so I think I'd see an animation from time to time, but can be cool, sure.
> Finally I have another request (I'm making assumptions about the languages you speak :D ):
> 
>     * @schneidr can you write down how you would write (fuzzy) time in German
> 
>     * @adocampo can you do the same for Spanish (Castilian, Catalan? I always forget what people around the world consider to be Spanish)
>       I speak Italian, so we should be able to make the code at least a bit less English-centric.
Castillian is the name of the spanish language. Even every single person who has spanish as mother language knows its "real" name is castillian, but the name "spanish" has  been popularized in the lasts decades. Besides, even they learn castillian at the school, most latam folks call at the language spanish. Now even in Spain many people call themselves "spanish" instead of castillian. So, summarizing, you can safely say people from Spain speaks Castillian and people from LATAM speaks Spanish, yet all of us really know the proper name is Castillian.
After all this diatribe, I can help with both castillian and catalan, yet catalan its a way more complex than spanish as it has 4 ways to say the time: watch time, bell time, traditional bell time, and international time, being the first more similar to english or any other romance language, international time is expressing the very exact time (like your first mockup), and bell time is the way we learn at school but almost nobody says on the real life, because you need to sum o substract quarter and minutes to say the time (video with captions in english here: https://www.youtube.com/watch?v=ecdd5eVRKu4)

Well, anyway, where do you want me to write down how do we say the time, right here?

adocampo commented:
Spanish
---
Name of the numbers:
1 / 13-> una (1 has male (uno) and female (una), the time it's female)
2  /14-> dos
3 / 15-> tres
4 / 16-> cuatro
5 / 17 -> cinco
6 / 18 -> seis
7 / 19 -> siete
8 / 20 -> ocho
9 / 21 -> nueve
10 / 22 -> diez
11 / 23 -> once
12 / 24 -> doce

Examples
14:00 --> Las dos en punto (13:00 or 01:00 would be  **La** una en punto as 1 is singular)
14:05 --> Las dos y cinco
14:10 --> Las dos y diez
14:15 --> Las dos y cuarto
14:20 --> Las dos y veinte
14:25 --> Las dos y veinticinco
14:30 --> Las dos y media
14:35 --> Las tres menos veinticinco (12:35 or 00:35 onwards would be **La** una menos veinticinco as 1 is singular)
14:40 --> Las tres menos veinte
14:45 --> Las tres menos cuarto
14:50 --> Las tres menos diez
14:55 --> Las tres menos cinco

Catalan
---
Name of the numbers:
1 / 13-> una (1 has male (un) and female (una), the time it's female)
2  /14-> dues (2 has male (dos) and female (dues), the time it's female)
3 / 15-> tres
4 / 16-> quatre
5 / 17 -> cinc
6 / 18 -> sis
7 / 19 -> set
8 / 20 -> vuit
9 / 21 -> nou
10 / 22 -> deu
11 / 23 -> onze
12 / 24 -> dotze

Examples (traditional bell system)
14:00 --> Les dues en punt (13:00 would be --> **La** una en punt as 1 is singular)
14:01-04 --> Les dues tocades (13:01-04 --> La una **tocada** (_tocada_ is the singular form of _tocades_)
14:05-06 --> Les dues ben tocades
14:07-08--> Mig quart de **_tres_** (in catalan, we say the hour we are going to, not the hour we come from)
14:09-11 --> Mig quart passat de tres
14:12-14 --> Mig quart ben passat de tres
14:15-16 --> Un quart de tres  (here we use "un", not "una", because "quarter" is male)
14:17-19 --> Un quart tocat de tres
14:20-21 --> Un quart ben tocat de tres
14:22-23 --> Un quart i mig de tres
14:24-26 --> Un quart i mig passat de tres
14:27-29 --> Un quart i mig ben passat de tres
14:30-31 --> Dos quarts de tres (here we use "dos", not "dues", because "quarter" is male)
14:32-34 --> Dos quarts tocats de tres
14:35-36 --> Dos quarts ben tocats de tres
14:37-38 --> Dos quarts i mig de tres
14:39-41 --> Dos quarts i mig passats de tres
14:42-44 --> Dos quarts i mig ben passats de tres
14:45-46 --> Tres quarts de tres
14:47-49 --> Tres quarts tocats de tres
14:50-51 --> Tres quarts ben tocats de tres
14:52-53 --> Tres quarts i mig de tres
14:54-56 --> Tres quarts i mig passats de tres
14:57-59 --> Tres quarts i mig ben passats de tres
Examples (clock system)
14:00 --> Les dues en punt (13:00 or 01:00 would be  **La** una en punt as 1 is singular)
14:05 --> Les dues i cinc
14:10 --> Les dues i deu
14:15 --> Les dues i quart
14:20 --> Les dues i vint
14:25 --> Les dues i vint-i-cinc
14:30 --> Les dues i mitja
14:35 --> Les dues menys vint-i-cinc (12:35 or 00:35 onwards would be **La** una menys vint-i-cinc as 1 is singular)
14:40 --> Les dues menys vint
14:45 --> Les dues menys quart
14:50 --> Les dues menys deu
14:55 --> Les dues menys cinc

dyamon commented:
Thanks for the very thorought explaination @adocampo 

I'll get to work!

schneidr commented:
Sorry for the delay, I was traveling and wasn't keen on writing this on the phone.

The numbers:

	 1 -> Ein/Eins
	 2 -> Zwei
	 3 -> Drei
	 4 -> Vier
	 5 -> Fünf
	 6 -> Sechs
	 7 -> Sieben
	 8 -> Acht
	 9 -> Neun
	10 -> Zehn
	11 -> Elf
	12 -> Zwölf

Additional words:

	to -> vor
	past -> nach
	quarter -> Viertel
	half -> halb

Examples:

	13:58-14:02 -> Zwei Uhr
	14:03-14:07 -> Fünf nach Zwei
	14:08-14:12 -> Zehn nach Zwei
	14:13-14:17 -> Viertel nach Zwei
	14:18-14:22 -> Zwanzig nach Zwei
	14:23-14:27 -> Fünf vor halb Drei
	14:28-14:32 -> Halb Drei
	14:33-14:37 -> Fünf nach halb Drei
	14:38-14:42 -> Zwanzig vor Drei
	14:43-14:47 -> Viertel vor Drei
	14:48-14:52 -> Zehn vor Drei
	14:53-14:57 -> Fünf vor Drei
	14:58-15:02 -> Drei Uhr
	
Special cases:

- Ein/Eins is a little tricky:
  - Ein Uhr
  - ... nach/vor/halb Eins
-  at midnight "Zwölf Uhr" could be replaced with "Mitternacht", but I don't think I would bother with it. And it would be only this one case, you wouldn't say "Fünf nach Mitternacht", you would still say "Fünf nach Zwölf".

I realize the umlauts are going to be a problem. The problem about such characters is known and I know it is being worked on, so for starters I think I would use the general replacements when the character is not available:

	ä -> ae
	ö -> oe
	ü -> ue
	
This can be changed later, when the characters are available.

Final note: I'm sure some of my countrymen are going to interject, saying "no, X is said differently here". How the time is said casually differs by dialect and region in Germany and other German speaking countries. Implementing all dialects would most probably not fit into the memory space of the PineTime, so I think it is best to stay with standard German.

ajack2001my commented:
Can this be put into v1.5.0?

schneidr commented:
> Can this be put into v1.5.0?

Doubtful that it is ready by then.

dyamon commented:
I have been testing what could be considered a "complete first version" of the watchface. I'll push this asap.
The process of reviewing and merging is not up to me, tho, so I cannot make any promises on v1.5.

Still missing:
- While code is written to easily integrate internationalisation, this is not a feature offered by InfiniTime atm,
- "Shake to show exact time" is not implemented (but still very much interested in experimenting with that)

kieranc commented:
I like this watchface a lot - I think I preferred the blue/white colours to the grey/white but that can easily be changed or user selected. The larger font looked better IMO. It could also be interesting to show icons (ble, battery etc) on long press or similar.

dyamon commented:
> I think I preferred the blue/white colours to the grey/white but that can easily be changed or user selected.

The color has always been grey/white (poor quality picture look blueish). I opened an issue to solve the problem of personal preferences regarding colors (#684). 

> The larger font looked better IMO. It could also be interesting to show icons (ble, battery etc) on long press or similar.

Definitely agree with you on everything. For fonts I guess we will have freedom when we will have access to the flash memory.

Itai-Nelken commented:
>when we will have access to the flash memory.

I think there already is.

dyamon commented:
Sure but afaik fonts are still embedded in the firmware right? I think most of the restriction we have with fonts will go away when we will have access to fonts stored in there.

dyamon commented:
I forgot to go through the process of generating Jetbrains Mono 42 again 🙄 Going to do it tomorrow

Riksu9000 reviewed (changes requested):
Looking really good!

Maybe you should update the font readme with instruction on generating the font as well. https://github.com/JF002/InfiniTime/blob/develop/src/displayapp/fonts/README.md

dyamon commented on src/displayapp/screens/settings/SettingWatchFace.cpp:
```diff
@@ -67,6 +67,15 @@ SettingWatchFace::SettingWatchFace(Pinetime::Applications::DisplayApp* app, Pine
   }
 
+  optionsTotal++;
+  cbOption[optionsTotal] = lv_checkbox_create(container1, nullptr);
```
It's declared of length 4. Am I missing something?

Riksu9000 replied:
The length is set to two here.
https://github.com/JF002/InfiniTime/blob/1c3c3c8db919c703bbd3f1be4ce9c6833237ea6f/src/displayapp/screens/settings/SettingWatchFace.h#L23

dyamon commented on src/displayapp/screens/WatchFaceFuzzy.cpp:
```diff
@@ -0,0 +1,200 @@
+  lv_obj_set_width(label_time, LV_HOR_RES);
+  lv_label_set_align(label_time, LV_LABEL_ALIGN_CENTER);
+  lv_label_set_recolor(label_time, true);
+  lv_obj_set_style_local_text_color(label_time, LV_LABEL_PART_MAIN, LV_STATE_DEFAULT, lv_color_hex(0x777777));
```
Speaking of default palettes 😛 Thanks, I didn't know these existed!

dyamon commented:
Thanks.

What are you suggesting for the font README? Is something like

```
If you are changing Jetbrains Mono 42 include at least the following symbols: ...
```

enough?

Riksu9000 commented:
> Thanks.
> 
> What are you suggesting for the font README? Is something like
> 
> ```
> If you are changing Jetbrains Mono 42 include at least the following symbols: ...
> ```
> 
> enough?

Currently there are no instructions on generating the 42px font, so to generate the font we have to guess the settings or figure them out from the source. I think it would be best to create separate sections for the 20px and 42px font which would have mostly the same format explaining each setting that we need to enter in the font converter.

Also the JetBrainsMono-Regular.ttf file should probably be added to the repo like the bold font used for the 20px font, so we don't need to go hunting for that separately as well. This would also make sure the version matches, because for example when you updated the font, numbers 2 and 3 changed as well, probably due to a newer version.

dyamon commented:
I updated the README for the font generation, and reworked it a bit. Let me know what you think.

Riksu9000 commented on src/displayapp/fonts/README.md:
```diff
@@ -6,47 +6,68 @@
+- Load `JetBrainsMono-Bold.tff` (use the file in this repo to ensure the version matches)
+- **Range:** `0x20-0x7f,0x410-0x44f` and add any additional glyphs
+- Include a second font
+- Load `FontAwesome5-Solid+Brands+Regular.woff` (use the file in this repo to ensure the version matches)
```
This file exists in a different place, so perhaps there should be a link or a path to the file. I actually didn't know about it before now.

dyamon replied:
I decided to specify the full path whenever it was relevant. This should be less confusing.

Riksu9000 commented on src/displayapp/screens/WatchFaceFuzzy.h:
```diff
@@ -0,0 +1,26 @@
+      private:
+        Controllers::DateTime& dateTimeController;
+        lv_task_t* taskRefresh;
+        lv_obj_t* label_time;
```
Variables should be camelCase.

dyamon replied:
Ops, wild copypasting :) Fixed

dyamon commented:
Pictures of the current build.

![fuzzy2](https://user-images.githubusercontent.com/6771224/134545001-75d031bc-d2ed-4a58-a8a6-4db86e955fe8.png)

Riksu9000 reviewed (commented):
Just a few minor things I noticed. Other than that everything looks good to me.

dyamon commented on src/displayapp/fonts/README.md:
````diff
@@ -6,47 +6,68 @@
+- **Range:** `0x20-0x7f,0x410-0x44f` and add any additional glyphs
+- Include a second font
+- Load `FontAwesome5-Solid+Brands+Regular.woff` (use the file in this repo to ensure the version matches)
+- **Range:**
+
+    ```
+    0xf293, 0xf294, 0xf244, 0xf240, 0xf242, 0xf243, 0xf241, 0xf54b, 0xf21e,
+    0xf1e6, 0xf54b, 0xf017, 0xf129, 0xf03a, 0xf185, 0xf560, 0xf001, 0xf3fd,
+    0xf069, 0xf1fc, 0xf45d, 0xf59f, 0xf5a0, 0xf029, 0xf027, 0xf028, 0xf6a9,
+    0xf04b, 0xf04c, 0xf048, 0xf051, 0xf095, 0xf3dd, 0xf04d, 0xf2f2, 0xf024,
+    0xf252, 0xf569, 0xf201, 0xf06e, 0xf015
+    ```
+
````
Yeah, every bullet list with an indented code block has the weird line spacing. I'd rather leave it like this because GitHub adds the handy "copy to clipboard" button to code block, which makes copy-pasting long lines like these less error prone.

dyamon commented on src/displayapp/fonts/README.md:
```diff
@@ -6,47 +6,68 @@
+- **Bpp:** 1 bit-per-pixel
+- Do *not* enable font compression or horizontal subpixel hinting
+- Do *not* enable use of glyph color
+- Load `JetBrainsMono-Bold.tff` (use the file in this repo to ensure the version matches)
```
Fixed

dyamon commented on src/displayapp/fonts/README.md:
```diff
@@ -6,47 +6,68 @@
+- **Bpp:** 1 bit-per-pixel
+- Do *not* enable font compression or horizontal subpixel hinting
+- Do *not* enable use of glyph color
+- Load `JetBrainsMono-Regular.tff` (use the file in this repo to ensure the version matches)
```
Fixed

schneidr commented:
> It could also be interesting to show icons (ble, battery etc)
Btw, my personal preference would be to hide icons by default and only show the Bluetooth icon when the watch is disconnected and the battery icon when the battery level is low.

adocampo commented:
> > It could also be interesting to show icons (ble, battery etc)
> 
> Btw, my personal preference would be to hide icons by default and only show the Bluetooth icon when the watch is disconnected and the battery icon when the battery level is low.

I would rather do as pebble did, fuzzy watchface is just that, without bloating it with anything else.
What it did, is when shaking the wrist, it showed for a few seconds another simple watchface with the time and date in digital format for those moments one needs to know the exact time with seconds. I think BT and battery information would fit there perfectly, and even step counter if you want.

OTOH, if https://github.com/JF002/InfiniTime/issues/659 and https://github.com/JF002/InfiniTime/issues/597 are both solved with showing a temporary watchface indicating BT connection is lost or battery is running out, there would be no need to bloat this fuzzy watchface, in my opinion.

dyamon commented:
I'm doing some experiments with wrist shaking atm. This would be compatible with what @adocampo is describing.

hubmartin commented:
@dyamon I just merged it and they looks awesome! Thanks
I just noticed that when Gadgetbridge updates the time (after reboot I had 0:00) then the text did not update itself, I had to go to menu and back. But at least digital watchface updates instantly.

dyamon commented:
That would be because the refresh time is set to 60s. The watchface is created before GB has time to sync the clock and you need to wait for 60s for the change to take place (or indeed "recreate" the watchface by going to a menu and back after GB synced the time). That said, this shouldn't be that noticeable anymore since:
- with the latest 1.5 we have clock persistence;
- I'm implementing "digital view on shake" and that needs an higher refresh rate anyway.

dyamon commented:
So this is a personal attempt at implementing a "show digital clock on shake" functionality. I'm getting a pretty consistent shake detection (even when my arm is not parallel to the ground) and haven't noticed false positives so far.

While this is working fine for me, feedback is very much appreciated.

_I can't really turn this PR into a draft anymore but it should be._

geekbozu commented:
> 
> 
> So this is a personal attempt at implementing a "show digital clock on shake" functionality. I'm getting a pretty consistent shake detection (even when my arm is not parallel to the ground) and haven't noticed false positives so far.
> 
> While this is working fine for me, feedback is very much appreciated. I'm also aware of #691 which implement a similar functionality, but apparently any sudden acceleration is considered a shake. We'll see what's the best solution to adopt.
> 
> _I can't really turn this PR into a draft anymore but it should be._

Yay multiple attempts to figure out what works best for us :D I think the major difference is you only use the Y axis, where I am using all 3 in some fashion. That said I am still heavily tweaking...I will locally give this a shot and see how it feels :)
I also add some averaging in which allows the shake momentum to build up over time, Still not sure if this is a good or bad idea yet. I personally like how it feels but my tastes may not be everyone! 
Keep up the awesome work :D

geekbozu commented:
Out of curiosity is there anyway to use the existing digital watch face instead of rolling your own?

dyamon commented:
Sorry it seems like I didn't reply to your previous comment :) 

> I think the major difference is you only use the Y axis, where I am using all 3 in some fashion.

I wonder whether using all 3 gives more accuracy for the movement I'm trying to capture...

> Out of curiosity is there anyway to use the existing digital watch face instead of rolling your own?

That would be indeed the ideal solution! I opted for the custom digital face because I'm not sure how to implement the "showing a watchface for x seconds" without modifying the target watchface. The only example we have is the notification app and that encodes both the timed and the standard version. Do you have any suggestion for this?

Avamander commented on src/displayapp/screens/WatchFaceFuzzy.cpp:
```diff
@@ -0,0 +1,258 @@
+    }
+
+    timeStr = timeSectors[sector];
+    if (timeStr.find("%1") != std::string::npos)
+      hours = (hours + 1) % 12;
```
Inconsistent brace usage.

dyamon replied:
Fixed

kieranc commented:
I still really like this watchface and would love for it to be merged, or failing that, used in my test builds, but the motion based switch to the digital watchface has been giving me false triggers, and it shares a lot of functionality with shake wake, which should be merged at some point..
How would you feel about slimming down this PR to a pure watchface, with a view to using ShakeWake's functionality to activate the digital display at a later date/once SW is merged? Geekbozu assures me the SW code can be used in this way.

dyamon commented:
I've been a bit busy but I'm planning to finish this over Christmas. I think the first step is to revert the latest changes that introduced the shaking functionality. And rebase that against master in order to merge the watchface.

Once that is done, I still would like to have the shake-to-digital, so I'll look into the new shake-to-wake code and see what can be reused. Also I originally came up with a custom digital face, but I should really use the default one instead as suggested above.

adocampo commented:
Glad to see this on the radar... I would love to have the fuzzy watchface. Can't wait for it!
The digital clock, when finally gets merged, will it be shown just a few seconds and then will automatically go to the fuzzy clock, or how do you think to work?

dyamon commented:
> The digital clock, when finally gets merged, will it be shown just a few seconds and then will automatically go to the fuzzy clock, or how do you think to work?

At the moment, it is doing that (similar to what a new notification does). The alternative (but this is something I thought about just now) is to have the shake toggle the face.

adocampo commented:
On pebble it was a short timeout (3'' to 5'' IIRC), but you were able to press the back button to "return" to the fuzzy watchface at any time.
Surely there will be needs for everything, and sometimes you will need to stay on the digital clock for a while (ie: when you need to count at seconds level), but I'm just fine with a timeout or pressing the back button. I'm not a huge fan of shaking on PineTime, it has a lot of false positives.

geekbozu commented:
> 

The shaking will get better soon, ShakeWake imo does it pretty well/Has a customizable threshold. Now under the hood there are some Fixes as well. We sometimes get bad data into the accelerometer causing it to act weird. Its on the radar just needs to get fixed :)

dyamon commented:
That's good news. So do you think the shake code is general enough to be reused as a, let's say, event with a callback? (Sorry haven't looked at it yet)

geekbozu commented:
> 

It could be extended to, Right now there is just a function to see if we are actively over the threshold or not.  The system task runs every 100ms or so so I just poll the value. Which should work well enough for you as well.
Obviously this can get extended in the future if needed. We ave lots of choices. :)

dyamon commented:
Oh ok, seems easy enough. Thanks!

@geekbozu I might ping you later on in the chat if you know something about timers and such for the "temporary digital" face.

kieranc commented:
> Oh ok, seems easy enough. Thanks!
> 
> @geekbozu I might ping you later on in the chat if you know something about timers and such for the "temporary digital" face.

It might be the situation has changed but I worked with geekbozu when trying to make a button disappear after a delay and we tried a few approaches using timers with no success. The eventual solution was simply to record the tick count and remove the button after a number of ticks. You can see the solution we used in PR #718

dyamon commented:
I think this is basically what I'm doing at the moment. I think notifications already use something different from tick counting so I might have a look at that.

dyamon commented:
So, I cleaned the code from custom shaking code for now and rebased everything against `develop`.

I'll turn this into an actual pull request so that it can be merged. I think a few people are already running this so it has been tested enough :)

Next I'll tackle the digital-on-shake feature, but it will be a different PR.

Itai-Nelken commented:
I found a bug:
When returning from the alarm app with a long button press, the display gets corrupted. This doesn't happen with other watchfaces.

dyamon commented:
Can confirm, corrupts the display for me as well, regardless of the app you are in. It seems like the screen gets shifted by a certain amount (possibly dependent on position of the app in the menu... this might be a hint). I'll have to look into it.

Any help is appreciated oc :)

Riksu9000 commented:
You'll need to add a `backgroundLabel` to force a full screen refresh. Check the other watchfaces. This will fix the issue.

dyamon commented:
That was quick, thanks. I remember removing that thinking "Why would I need that anyway?" and that's the reason :)

dyamon commented:
Fixed it in the latest commit.

dyamon commented:
I merged some changes that add the "digital on shake" feature. I have been testing it for a week or so and it seems pretty reliable to me (assuming you set a good threshold for the wakesense).

- I would appreciate some feedback on the code. I changed some common code and now it should be easier to create "preview" version of an app.
- This feature doesn't play too well with the shake2wake feature since shaking with the screen turned off most likely triggers both features at once. I tried a few things to avoid this but it didn't work and didn't really want to add custom hacks. If this is something that really bothers people we can try and think of a solution.

dyamon commented:
Hey, does anyone have any more feedback on this? I have been running it for a while, and afaik someone else is as well and have no problem.

It would be nice to get this finally merged after a rebase.

kieranc commented:
I think it's a great watchface and I'd love to see it merged, do you have figures on how much additional flash space it uses? We're pretty tight on space currently and I'm worried the additional font will push us even closer to the storage limit.
What we really need is to get fonts moved to SPI flash but that's a big undertaking and everyone is busy currently :-( It's not just moving them but also implementing a way to install them to flash alongside DFU updates, presumably using BLE FS.

JF002 commented:
Exactly as Kieran said : RAM and flash memories are pretty packed right now, and I refrain from merging features that needs too much memory right now (and fonts takes a lot of space, indeed). 

It's frustrating as there are a few nice watchfaces that are waiting to be merged, but my goal is to avoid hitting the memory limit, which would force us to remove features from the firmware.

dyamon commented:
Hey, I'm joining you guys on the being busy train, so I'm not up to date with the latest news on the development.
I understand the situation. I'm around anyway (playing with InfiniTime is something I always go back to) so ping me if/when you decide to merge this (hopefully with the changes on fonts handling).

ajack2001my commented:
I think the InfiniTime team needs to have a mechanism where devs can write watchfaces and download them separate from the firmware and store in watch storage space and not in running memory like what I understand is happening at the moment. Please correct me if I'm wrong here...

kieranc commented:
> Hey, I'm joining you guys on the being busy train, so I'm not up to date with the latest news on the development. I understand the situation. I'm around anyway (playing with InfiniTime is something I always go back to) so ping me if/when you decide to merge this (hopefully with the changes on fonts handling).
Thanks for your contribution and understanding :-)


> I think the InfiniTime team needs to have a mechanism where devs can write watchfaces and download them separate from the firmware and store in watch storage space and not in running memory like what I understand is happening at the moment. Please correct me if I'm wrong here...
You're correct, this is something we'd love to implement but it is also a pretty big goal. I think once we have fonts and other large assets moved to the external flash there should be space for quite a few watchfaces, until we have dynamic loading ready.

ajack2001my commented:
> You're correct, this is something we'd love to implement but it is also a pretty big goal. I think once we have fonts and other large assets moved to the external flash there should be space for quite a few watchfaces, until we have dynamic loading ready.

Not forgetting an app store for other apps and watchfaces; or a payment mechanism... :)

ajack2001my commented:
@JF002 when will such discussions take place?

JF002 commented:
> @JF002 when will such discussions take place?

Discussions about dynamic app loading and storage of big assets? We often talk about this in the chat room and in some PR/issues when the topic is addressed. There's no "centralized" place to talk about this, but feel free to open an issue to open the discussion.

In my opinion, we all agree we need to work on that kind of functionality, but strangely enough, a lot of people talk about this but unfortunately, very few experiment and try to find actual solution to implement them ;-) 
This is in my TODO list but... there are a lot of things in my TODO list and very little time to work on them :)

calhix commented:
Would someone be able to upload a build with this watchface for 1.10?

dyamon commented:
Sure thing, I could be able to resolve conflicts and post a build by the end of the week. Atm I'm travelling and don't have my PineTime with me, to test if everything works as intended, maybe someone can help with that.

adocampo commented:
Too bad my PineTime isn't charging anymore :'( ... I had to buy a Pebble Time Steel on eBay to enjoy a fuzzy watch face. 
Anyway, I hope this can see the light someday. :)

dyamon commented:
@calhix
Hey here is a build (on top of master) including the fuzzy watchface. Flashed right now and everything seems to work. As a bonus it includes also the Infineat watchface #1024 recently merged (it has some visual bugs tho, have to investigate).

[pinetime-mcuboot-app-dfu-1.10.0.zip](https://github.com/InfiniTimeOrg/InfiniTime/files/9673932/pinetime-mcuboot-app-dfu-1.10.0.zip)

@adocampo 
sorry to hear that :( but you never know, Santa might bring you a PineTime this year since Pine64 is selling in the EU as well.

@JF002
Just saw that we can now load resources from memory. Adapting the code to this new feature would be enough to merge this PR?

ajack2001my commented:
@dyamon the implementation of the fuzzy watchface is buggy, it randomly switches from fuzzy to digital... Hope you can fix this since this is only a test build and not official... :'(

FintasticMan commented:
That isn't a bug, that is the feature that when you shake the watch it switches back to the digital watch face. I personally don't like the feature, but it is intended behaviour.

dyamon commented:
> @dyamon the implementation of the fuzzy watchface is buggy, it randomly switches from fuzzy to digital... Hope you can fix this since this is only a test build and not official... :'(

About this, what you probably want to do is to perform a shake calibration (settings (3rd page) > Shake Calib.) I think this is set to the minimum as a default, which leads to this "random" activation of the digital preview.

ajack2001my commented:
Thanks for the update of the fuzzy watchface @FintasticMan @dyamon... Though I only notice it happen with that watchface and not the others. I will recalibrate as suggested, but is there a way to just turn that feature off? In any case, thanks again guys! Appreciate your time and efforts into this...

dyamon commented:
The feature is indeed baked into **this** watchface.
I'm not a fan of making everything a setting (it's a type of feature creep. Also possibly relevant [xkcd strip](https://xkcd.com/1172/)), but I can look into it. A lot of watchfaces seem to have contextual menu atm.

Temporary solution: if you are not using the shaking feature for anything else (e.g., shake to wake or similar) you could crank up the shake threshold so you won't ever trigger the digital clock.

calhix commented:
That's what I ended up doing so I never see the digital face anymore, I do think a better solution would be to tap/double-tap the screen to temporarily show the steps and date on the fuzzy watchface itself.

dyamon commented:
Personally I occasionally want to know the exact time, but maybe that's just me? I guess we can find a solution/compromise.

adocampo commented:
> Personally I occasionally want to know the exact time, but maybe that's just me? I guess we can find a solution/compromise.

I'm wiht you. It's interesting sometimes to know the exact time (i.e. when cooking, counting something, etc)

calhix commented:
Just wanted to say you did a great job on this. Would be great if we could have this ported to 1.11 resources (like with digistyle in #1116) so that it could possibly be included in the next update.

calhix commented:
> Personally I occasionally want to know the exact time, but maybe that's just me? I guess we can find a solution/compromise.

Perhaps a possible solution is for the watchface to become sort of like #1306 when shaking the wrist except in your style? May be a bit of work though.

schneidr commented:
> Just wanted to say you did a great job on this. Would be great if we could have this ported to 1.11 resources (like with digistyle in #1116) so that it could possibly be included in the next update.

I was thinking maybe it's even possible to store all language specific resources that way, making it possible to add additional language (with language specific fonts for characters now missing) without changing the watchface itself.

dyamon commented:
> Just wanted to say you did a great job on this. Would be great if we could have this ported to 1.11 resources (like with digistyle in #1116) so that it could possibly be included in the next update.

Thanks, I have been updating the PR every now and then when I update my PineTime.
When the main team decides to merge the feature I'll gladly port it to the next major release.

> Optimally to fit the style of this watch face would be to shake or do some gesture to have the exact time written out as well (like "seven after eleven" instead of "five after eleven." Would be a lot of work though.

Incidentally, this was the initial behavior (see first screenshot) when I didn't really get what a fuzzy watch was :D 

> I was thinking maybe it's even possible to store all language specific resources that way, making it possible to add additional language (with language specific fonts for characters now missing) without changing the watchface itself.

I'd definitely go in that direction but wait to see if there is any work on internationalization at a more general level, otherwise we duplicate work for every app that wants to support multiple languages.

Zandengoff commented:
@dyamon Do you know what failed in the checks? Seems like with a small tweak this could be merged.

dyamon commented:
Hey @Zandengoff, I haven't being following the project lately but this is basically a rebase on the new changes.
Before diving into the rebase, tho, I'd like a confirmation that this is going to be merged, otherwise I'll keep running after changes.

kieranc commented:
I'd love yo see this merger, I haven't looked at the debased version yet but I will. Does it have the fonts stored externally?

kieranc commented:
So the font is still the internal 40px with additional ranges. You could use save some internal flash space and make it look nicer by putting the 42px font on ext flash. As far as the shake/digital switch, I'm not a big fan, but maybe an easy option would be to bind long press to display the digital time temporarily?

zyphlar commented:
Would love for this to be an option in the main distribution (I thought it was when I bought a PineTime lol)

zyphlar commented:
I hacked together an adaptation of this suited for the current state of the code that adds the fuzzy language as an option to the digital watchface instead of making a whole new watchface (which I couldn't get to fit on the device lol) -- might serve as a basis for someone else to hack on since previous implementations are based on code that's changed substantially

https://github.com/InfiniTimeOrg/InfiniTime/compare/main...zyphlar:InfiniTime:wb/fuzzy
